### PR TITLE
[Kernel][Model] PagedAttention: Support custom attention bias for T5 model (1/2)

### PR DESCRIPTION
**Pull request by NickLucche**
This contribution is part of 2 PRs attempting to add support for the T5 model.
Based on the great work done in https://github.com/vllm-project/vllm/issues/7366#user-content-support-custom-attention-bias and https://github.com/vllm-project/vllm/pull/3117.

It implements:

$$
Attention(Q, K, V) = \text{softmax}\left(\frac{QK^\top + A}{\sqrt{d_k}}\right)V
$$

In particular, the following aims to be a more generalized addition (not tied to T5), enabling any 
custom attention bias in the PagedAttention kernel.

While I am aware this introduces a fully materialized matrix (akin to what's done in xformers), 
currently padded to `max_seq_len`, I believe the flexibility brought by allowing any custom bias
would still pay for itself when providing initial compatibility for yet unsupported models (such as T5). 
I'd be happy to focus on performance optimization later (eg variable len bias matrix or flashattn-like IO optimization)
or even have yet another model-specific feature in the kernel for T5 to compute relative positional embeddings.

TODO work left for future contribs (ie extend to all platforms):
 - ROCm attn bias support
 - CPU attn bias support
 - ..?

### CODE REVIEW

**Comment by NickLucche** [comment]
Thanks for the input @sfc-gh-phalama, personally I'd rather not have the attn bias implementation depend or be tied to T5/relative positional encoding mechanisms, otherwise we could probably implement the inline formula.

**Comment by NickLucche** [comment]
Fixed the kernel tests failing, not sure how related the rest of errors may be tbh..
https://buildkite.com/vllm/ci/builds/12022#01946b42-da66-4512-919d-c2115515eb6d/147-444

**Comment by joerunde** [comment]
@NickLucche the gpu count issue might be fixed by this PR: https://github.com/vllm-project/vllm/pull/12111

**Comment by NickLucche** [comment]
CI failures still look unrelated to me :/

**Comment by NickLucche** [comment]
I believe this PR could also be useful in places like https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/mllama.py#L863, where we can pass in the custom attention mask as bias term A, while getting rid of the logic to handle kv cache in the model.

**Comment by NickLucche** [comment]
Rebased

**Review by hmellor — COMMENTED** [review]
I'm partway through all the files but I have 2 initial questions:
- Do the `attn_bias` arguments need to be added to the other attention backends?
- Is the new tensor materialised even if the attention bias is not used?

**Comment by NickLucche** [comment]
Hey thanks for the review!
> Do the attn_bias arguments need to be added to the other attention backends?

Initial effort was planned only for the pagedattn backend, so that we could run at least with xformers+pagedattn. FlashAttn would require a separate PR on base repo or on vllm maintained fork, so that requires more involvement.

> Is the new tensor materialised even if the attention bias is not used?
Nope attn_bias it's an optional pointer, when the pointer is null it's a noop.
